src/libraries/System.Formats.Tar/src/System/Formats/Tar/SubReadStream.cs

```diff
@@ -0,0 +1,174 @@
+﻿// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics;
+using System.IO;
+using System.Runtime.InteropServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Formats.Tar
+{
+    // Stream that allows wrapping a super stream and specify the lower and upper limits that can be read from it.
+    // It is meant to be used when the super stream is unseekable.
+    // Does not support writing.
+    internal class SubReadStream : Stream
+    {
+        protected readonly long _startInSuperStream;
+        protected long _positionInSuperStream;
+        protected readonly long _endInSuperStream;
+        protected readonly Stream _superStream;
+        protected bool _isDisposed;
+
+        public SubReadStream(Stream superStream, long startPosition, long maxLength)
+        {
+            if (!superStream.CanRead)
+            {
+                throw new NotSupportedException(SR.IO_NotSupported_UnreadableStream);
+            }
+            _startInSuperStream = startPosition;
+            _positionInSuperStream = startPosition;
+            _endInSuperStream = startPosition + maxLength;
+            _superStream = superStream;
+            _isDisposed = false;
+        }
+
+        public override long Length
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return _endInSuperStream - _startInSuperStream;
+            }
+        }
+
+        public override long Position
+        {
+            get
+            {
+                ThrowIfDisposed();
+                return _positionInSuperStream - _startInSuperStream;
+            }
+            set
+            {
+                ThrowIfDisposed();
+                throw new NotSupportedException(SR.IO_NotSupported_UnseekableStream);
+            }
+        }
+
+        public override bool CanRead => !_isDisposed;
+
+        public override bool CanSeek => false;
+
+        public override bool CanWrite => false;
+
+        internal bool WasStreamAdvanced
+        {
+            get => _positionInSuperStream >= _endInSuperStream;
+            set
+            {
+                if (value)
+                {
+                    _positionInSuperStream = _endInSuperStream + 1;
+                }
+            }
+        }
+
+        protected void ThrowIfDisposed()
+        {
+            if (_isDisposed)
+            {
+                throw new ObjectDisposedException(GetType().ToString(), SR.IO_StreamDisposed);
+            }
+        }
+
+        private void ThrowIfBeyondEndOfStream()
+        {
+            if (_positionInSuperStream >= _endInSuperStream)
```

Comment:
Yes: If the user is working with an unseekable stream, opens it with a TarReader, gets one entry, stores it in a variable, then gets the next entry, then attempts to retrieve the DataStream of the first entry. Example:

```cs
void Example(Stream unseekableStream)
{
    using TarReader reader = new(unseekableStream);
    TarEntry entry1 = reader.GetNextEntry(copyData: false); // Assuming it's a regular file entry, the user wasn't careful to store the data in an new memory stream
    TarEntry entry2 = reader.GetNextEntry();
    entry1.DataStream.Read( .... ); // This would throw EndOfStreamException due to attempting to read the super stream
}
```

I wrote a unit test to verify this:

https://github.com/carlossanlop/runtime/blob/84ecc16df28bbac7224b3274c4666f52a56cd453/src/libraries/System.Formats.Tar/tests/TarReader/TarReader.GetNextEntry.Tests.cs#L103